### PR TITLE
Refactor quote matching internal logic

### PR DESCRIPTION
**Pull request by nicolasstucki**
Aimed to simplify the logic exposed by the CompilerInterface

### CODE REVIEW

**Comment by nicolasstucki** [comment]
@b-studios this may be of interest for you

**Comment by b-studios** [comment]
Yes, thanks! That goes into the same direction as #9872 

https://github.com/lampepfl/dotty/blob/4dd7062ff8c03ac4d76c9c49a11361dbab49c16b/library/src-bootstrapped/scala/quoted/Scope.scala#L25-L26

File: library/src/scala/internal/tasty/CompilerInterface.scala [hunk 1]
```diff
@@ -21,26 +21,45 @@ trait CompilerInterface { self: scala.tasty.Reflection =>
...
 
-
-  def Constraints_context[T]: scala.quoted.QuoteContext
-  def Constraints_add(syms: List[Symbol]): Boolean
-  def Constraints_approximation(sym: Symbol, fromBelow: Boolean): Type
+  /** Pattern matches an the scrutinee against the pattern and returns a tuple
+   *  with the matched holes if successful.
+   *
+   *  Examples:
+   *    - `termMatch(< f(0, myInt) >, < f(0, myInt) >, false)`
```
Comment (liufengyun) [inline library/src/scala/internal/tasty/CompilerInterface.scala:28]:
```suggestion
   *    - `termMatch(< f(0, myInt) >, < f(0, myInt) >)`
```

File: library/src/scala/internal/tasty/CompilerInterface.scala [hunk 2]
```diff
@@ -21,26 +21,45 @@ trait CompilerInterface { self: scala.tasty.Reflection =>
...
+   *       will return `Some(())` (where `()` is a tuple of arity 0)
+   *    - `termMatch(< f(0, myInt) >, < f(patternHole[Int], patternHole[Int]) >, false)`
```
Comment (liufengyun) [inline library/src/scala/internal/tasty/CompilerInterface.scala:30]:
```suggestion
   *    - `termMatch(< f(0, myInt) >, < f(patternHole[Int], patternHole[Int]) >)`
```

File: library/src/scala/internal/tasty/CompilerInterface.scala [hunk 3]
```diff
@@ -21,26 +21,45 @@ trait CompilerInterface { self: scala.tasty.Reflection =>
...
+   *       will return `Some(Tuple2('{0}, '{ myInt }))`
+   *    - `termMatch(< f(0, "abc") >, < f(0, patternHole[Int]) >, false)`
```
Comment (liufengyun) [inline library/src/scala/internal/tasty/CompilerInterface.scala:32]:
```suggestion
   *    - `termMatch(< f(0, "abc") >, < f(0, patternHole[Int]) >)`
```

File: library/src/scala/internal/tasty/CompilerInterface.scala [hunk 4]
```diff
@@ -21,26 +21,45 @@ trait CompilerInterface { self: scala.tasty.Reflection =>
    */
   def unpickleType(repr: Unpickler.PickledQuote, args: Unpickler.PickledArgs): TypeTree
 
-
-  def Constraints_context[T]: scala.quoted.QuoteContext
-  def Constraints_add(syms: List[Symbol]): Boolean
-  def Constraints_approximation(sym: Symbol, fromBelow: Boolean): Type
+  /** Pattern matches an the scrutinee against the pattern and returns a tuple
```
Comment (liufengyun) [inline library/src/scala/internal/tasty/CompilerInterface.scala:24]:
```suggestion
  /** Pattern matches the scrutinee against the pattern and returns a tuple
```

File: library/src/scala/internal/tasty/CompilerInterface.scala [hunk 5]
```diff
@@ -21,26 +21,45 @@ trait CompilerInterface { self: scala.tasty.Reflection =>
...
+   *    - scala.internal.Quoted.patternHole[T]: hole that matches an expression `x` of type `Expr[U]`
+   *                                            if `U <:< T` and returns `x` as part of the match.
+   *
+   *  @param scrutinee `Term` on which we are pattern matching
+   *  @param pattern `Term` containing the pattern tree
+   *  @return None if it did not match, `Some(tup)` if it matched where `tup` contains `Term``
+   */
+  def termMatch(scrutinee: Term, pattern: Term): Option[Tuple]
+
+  /** Pattern matches an the scrutineeType against the patternType and returns a tuple
```
Comment (liufengyun) [inline library/src/scala/internal/tasty/CompilerInterface.scala:45]:
```suggestion
  /** Pattern matches the scrutineeType against the patternType and returns a tuple
```

**Review by liufengyun — APPROVED** [review]
LGTM

**Pull request merged**